File: app/src/main/java/org/digitalcampus/oppia/utils/custom_prefs/TimePreferenceDialogFragmentCompat.java
package org.digitalcampus.oppia.utils.custom_prefs;


import android.content.Context;
import android.view.View;
import android.widget.TimePicker;

import androidx.preference.DialogPreference;
import androidx.preference.Preference;
import androidx.preference.PreferenceDialogFragmentCompat;

public class TimePreferenceDialogFragmentCompat extends PreferenceDialogFragmentCompat implements DialogPreference.TargetFragment {
    TimePicker timePicker = null;

    @Override
    protected View onCreateDialogView(Context context) {
        timePicker = new TimePicker(context);
        return (timePicker);
    }

    @Override
    protected void onBindDialogView(View v) {
        super.onBindDialogView(v);
        timePicker.setIs24HourView(true);
        TimePreference pref = (TimePreference) getPreference();
        timePicker.setCurrentHour(pref.hour);
        timePicker.setCurrentMinute(pref.minute);
    }

    @Override
    public void onDialogClosed(boolean positiveResult) {
        if (positiveResult) {
            TimePreference pref = (TimePreference) getPreference();
            pref.hour = timePicker.getCurrentHour();
            pref.minute = timePicker.getCurrentMinute();

            String value = TimePreference.timeToString(pref.hour, pref.minute);
            if (pref.callChangeListener(value)) pref.persistStringValue(value);
        }
    }

    @Override
    public Preference findPreference(CharSequence charSequence) {
        return getPreference();
    }
}

File: app/src/main/java/org/digitalcampus/oppia/utils/custom_prefs/TimePreferenceDialogFragmentCompat.kt
package org.digitalcampus.oppia.utils.custom_prefs

import android.content.Context
import android.view.View
import android.widget.TimePicker
import androidx.preference.DialogPreference.TargetFragment
import androidx.preference.Preference
import androidx.preference.PreferenceDialogFragmentCompat
import org.digitalcampus.oppia.utils.custom_prefs.TimePreference.Companion.timeToString

class TimePreferenceDialogFragmentCompat : PreferenceDialogFragmentCompat(), TargetFragment {
    var timePicker: TimePicker? = null
    override fun onCreateDialogView(context: Context): View? {
        timePicker = TimePicker(context)
        return timePicker
    }

    override fun onBindDialogView(v: View) {
        super.onBindDialogView(v)
        timePicker?.setIs24HourView(true)
        val pref = preference as TimePreference
        timePicker?.currentHour = pref.hour
        timePicker?.currentMinute = pref.minute
    }

    override fun onDialogClosed(positiveResult: Boolean) {
        if (positiveResult) {
            val pref = preference as TimePreference
            pref.hour = timePicker!!.currentHour
            pref.minute = timePicker!!.currentMinute
            val value = timeToString(pref.hour, pref.minute)
            if (pref.callChangeListener(value)) pref.persistStringValue(value)
        }
    }

    override fun <T : Preference?> findPreference(key: CharSequence): T? {
        return preference as T?
    }
}